klib/mbedtls.c

```diff
@@ -0,0 +1,301 @@
+#include <klib.h>
+#include <runtime.h>
+#include <lwip.h>
+
+#define _RUNTIME_H_ /* guard against double inclusion of runtime.h */
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/entropy.h>
+#include <mbedtls/ssl.h>
+
+declare_closure_struct(1, 1, buffer_handler, tls_conn_handler,
+                       struct tls_conn *, conn,
+                       buffer_handler, out);
+declare_closure_struct(1, 1, status, tls_in_handler,
+                       struct tls_conn *, conn,
+                       buffer, b);
+declare_closure_struct(1, 1, status, tls_out_handler,
+                       struct tls_conn *, conn,
+                       buffer, b);
+
+typedef struct tls_conn {
+    mbedtls_ssl_context ssl;
+    closure_struct(tls_conn_handler, ch);
+    closure_struct(tls_in_handler, in);
+    buffer_handler out;
+    connection_handler app_ch;
+    buffer_handler app_in;
+    closure_struct(tls_out_handler, app_out);
+    buffer incoming, outgoing;
+    enum {
+        tls_handshake,
+        tls_open,
+        tls_closing,
+    } state;
+} *tls_conn;
+
+struct kern_funcs kern_funcs;
+
+static struct {
+    mbedtls_ssl_config conf;
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_entropy_context entropy;
+    mbedtls_x509_crt cacert;
+    status (*direct_connect)(heap h, ip_addr_t *addr, u16 port, connection_handler ch);
+    void (*destruct_tuple)(tuple t, boolean recursive);
+    u64 (*random_buffer)(buffer b);
+    struct tm *(*gmtime_r)(const long *timep, struct tm *result);
+    heap h;
+} tls;
+
+static void tls_close(tls_conn conn)
+{
+    if (conn->out)
+        apply(conn->out, 0);    /* close underlying TCP connection */
+    if (conn->app_in)
+        apply(conn->app_in, 0); /* notify connection shutdown to application layer */
+    else if (conn->app_ch)
+        apply(conn->app_ch, 0); /* notify connection failure to application layer */
+    mbedtls_ssl_free(&conn->ssl);
+    deallocate(tls.h, conn, sizeof(*conn));
```

Comment:
There was indeed a chance that conn->in is invoked after this deallocate, because the null buffer passed to conn->out to close the connection might be queued and processed asynchronously. Since there was no way for a direct connection client to know when a connection has been shut down, I modified the net direct code (and for consistency also the tls code) so that the receive handler is always invoked (with a null buffer) to notify a connection shutdown to the upper layer, regardless of which end of the connection initiated the shutdown. And the tls code now ensures that the tls_conn struct is deallocated after connection shutdown.